static/client.js

```diff
@@ -0,0 +1,252 @@
+ * - or parcel (https://parceljs.org).
+ * 
+ * DOM APIs are employed directly for rendering, which can be awkward.
+ * Rendering frameworks can help define more expressive dynamic trees:
+ * 
+ * - React (https://reactjs.org) / Preact (https://preactjs.com)
+ * - Vue (https://vuejs.org)
+ */
+
+/** Display a timestamp header after N seconds between messages. */
+var MIN_TIMESTAMP_HEADER_DELTA = 15 * 60;
+
+/** Ping every N seconds. */
+var PING_FREQ = 30;
+
+var connection;
+
+var session = {
+  username: null,
+  token: null, // authenticates username
+  lastMessageTimestamp: 0, // last received timestamp for header display deltas
+};
+
+/** Sends `message` as serialized JSON to WS host, including auth fields. */
+function sendJson(message) {
+  connection.send(JSON.stringify(Object.assign({
+    token: session.token,
+    username: session.username,
+  }, message)));
+}
+
+/** Converts UNIX timestamp to ISO-formatted string. */
+function unixTimeToIso(timestamp) {
+  var date = new Date();
+  date.setTime(timestamp * 1000);
+  return date.toUTCString();
+}
+
+/** Renders a received broadcast message. */
+function renderMessage(message) {
+  if (message.join) {
+    return renderActivityMessage(message.username, message.timestamp, 'joined');
+  }
+
+  if (message.leave) {
+    return renderActivityMessage(message.username, message.timestamp, 'left');
+  }
+
+  // Text content.
+  var container = renderMessageContainer(message.timestamp);
+  container.classList.add('content');
+  container.appendChild(renderBubble(message.username, message.content));
+  return container;
+}
+
+/**
+ * Renders a container for a received broadcast message.
+ * If `MIN_TIMESTAMP_HEADER_DELTA` seconds have passed since the last received
+ * message, the first child of the element will be a current timestamp header.
+ * Otherwise empty.
+ */
+function renderMessageContainer(timestamp) {
+  var container = document.createElement('li');
+  var timestampOptional = maybeRenderTimestampHeader(timestamp);
+
+  if (timestampOptional) {
+    container.appendChild(timestampOptional);
+  }
+
+  return container;
+}
+
+/**
+ * If `MIN_TIMESTAMP_HEADER_DELTA` seconds have passed since the last received
+ * message, a timestamp header is returned. Otherwise `null`.
+ */
+function maybeRenderTimestampHeader(timestamp) {
+  var delta = timestamp - session.lastMessageTimestamp;
+
+  session.lastMessageTimestamp = timestamp;
+
+  if (delta < MIN_TIMESTAMP_HEADER_DELTA) {
+    return null;
+  }
+
+  var header = document.createElement('div');
+  header.classList.add('meta');
+  header.classList.add('timestamp');
+  header.textContent = unixTimeToIso(timestamp);
+  return header;
+}
+
+/**
+ * Renders an activity message in format `${username} ${verbPastTense}.`, e.g.
+ * "alan has joined".
+ * Includes timestamp header if `MIN_TIMESTAMP_HEADER_DELTA` seconds have passed
+ * since the last received message.
+ */
+function renderActivityMessage(username, timestamp, verbPastTense) {
```

Comment:
If this accepted the `message` object, it could keep the localization concern local, and it could always be a pass thru so the calling function could avoid additional work based on return.

This is not needed for this example, just noticed while following the code flow.